### PR TITLE
MODPUBSUB-243 Add cache tenantToken invalidation

### PR DESCRIPTION
**Pull request by roman-barannyk**

### CODE REVIEW

File: mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java [hunk 1]
```diff
@@ -265,8 +232,114 @@ public void shouldSendRequestAndRetry(TestContext context) {
...
 
+  @Test
+  public void shouldInvalidateCacheBeforeRetryIfBadRequest(TestContext context) {
+    WireMock.stubFor(WireMock.post(CALLBACK_ADDRESS).willReturn(WireMock.badRequest()));
+    checkThatInvalidateTokenWasInvoked(context);
+  }
+
+  @Test
+  public void shouldInvalidateCacheBeforeRetryIfForbidden(TestContext context) {
+    WireMock.stubFor(WireMock.post(CALLBACK_ADDRESS).willReturn(WireMock.badRequest()));
```
Comment (roman-barannyk) [inline mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java:248]:
```suggestion
    WireMock.stubFor(WireMock.post(CALLBACK_ADDRESS).willReturn(WireMock.forbidden));
```

File: mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java [hunk 2]
```diff
@@ -265,8 +232,114 @@ public void shouldSendRequestAndRetry(TestContext context) {
...
+  }
+
+  @Test
+  public void shouldInvalidateCacheBeforeRetryIfBadRequestEntity(TestContext context) {
+    WireMock.stubFor(WireMock.post(CALLBACK_ADDRESS).willReturn(WireMock.badRequestEntity()));
+    checkThatInvalidateTokenWasInvoked(context);
+  }
+
+  @Test
+  public void shouldNotInvalidateCacheBeforeRetryIfNotContent(TestContext context) {
```
Comment (roman-barannyk) [inline mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java:259]:
```suggestion
  public void shouldNotInvalidateCacheBeforeRetryIfNoContent(TestContext context) {
```

File: mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java [hunk 3]
```diff
@@ -265,8 +232,114 @@ public void shouldSendRequestAndRetry(TestContext context) {
...
+    WireMock.stubFor(WireMock.post(CALLBACK_ADDRESS).willReturn(WireMock.noContent()));
+    checkThatInvalidateTokenWasNotInvoked(context);
+  }
+
+  @Test
+  public void shouldNotInvalidateCacheBeforeRetryIfOk(TestContext context) {
```
Comment (roman-barannyk) [inline mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java:265]:
```suggestion
  public void shouldNotInvalidateCacheIfOk(TestContext context) {
```

File: mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java [hunk 4]
```diff
@@ -265,8 +234,114 @@ public void shouldSendRequestAndRetry(TestContext context) {
...
+
+    consumerService.deliverEvent(event, params).onComplete(ar -> {
+      assertTrue(ar.succeeded());
+      verify(securityManager, never()).invalidateToken(TENANT);
+      verify(cache, never()).invalidateToken(TENANT);
+      async.complete();
+    });
+  }
+
+  private Set<MessagingModule> prepareMessagingModules() {
```
Comment (alexanderkurash) [inline mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java:312]: prepare -> build

File: mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java [hunk 5]
```diff
@@ -265,8 +234,114 @@ public void shouldSendRequestAndRetry(TestContext context) {
...
+      .withModuleId("mod-source-record-storage-1.0.0")
+      .withTenantId(TENANT)
+      .withModuleRole(MessagingModule.ModuleRole.SUBSCRIBER)
+      .withActivated(true)
+      .withSubscriberCallback(CALLBACK_ADDRESS));
+
+    return messagingModules;
+  }
+
+  private Event prepareEvent() {
```
Comment (alexanderkurash) [inline mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java:326]: prepare -> build

File: mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java [hunk 6]
```diff
@@ -265,8 +234,114 @@ public void shouldSendRequestAndRetry(TestContext context) {
...
+      .withId(UUID.randomUUID().toString())
+      .withEventType(EVENT_TYPE)
+      .withEventMetadata(new EventMetadata()
+        .withTenantId(TENANT)
+        .withEventTTL(30)
+        .withPublishedBy("mod-very-important-1.0.0"));
+  }
+
+  @NotNull
+  private OkapiConnectionParams prepareOkapiConnectionParams() {
```
Comment (alexanderkurash) [inline mod-pubsub-server/src/test/java/org/folio/services/impl/ConsumerServiceUnitTest.java:337]: prepare -> build

File: mod-pubsub-server/src/main/java/org/folio/services/impl/KafkaConsumerServiceImpl.java [hunk 7]
```diff
@@ -156,6 +156,10 @@ protected Handler<AsyncResult<RestUtil.WrappedResponse>> getEventDeliveredHandle
           event.getEventType(), event.getId(), subscriber.getSubscriberCallback(), ar.result().getCode(), ar.result().getResponse().statusMessage());
         LOGGER.error(error);
         auditService.saveAuditMessage(constructJsonAuditMessage(event, tenantId, AuditMessage.State.REJECTED, error));
+        if (ar.result().getCode() >= 400 && ar.result().getCode() < 500) {
```
Comment (alexanderkurash) [inline mod-pubsub-server/src/main/java/org/folio/services/impl/KafkaConsumerServiceImpl.java:159]: else {
  var ...
  if {

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**